### PR TITLE
[UWP] Added Windows Platform Specific to avoid set the default image directory

### LINKED ISSUE #2235: Improved image organization
## Summary
I would like to propose an improved image asset organization functionality for Xamarin.Forms 3.0.0, especially for UWP, but optionally on other OS as well.

### Default UWP image location change

Xamarin.Forms developers targeting UWP are currently forced to add their image assets directly in the root folder of the UWP project. This quickly becomes very unpleasant experience with tens of images right in the root folder next to files that the developer needs to refer to quickly like `App.xaml` of `Package.appxmanifest` and makes the project look unorganized. While using a shared project to have the images may provide a partial relief of this problem, it is still not a very nice user experience.

Usually in UWP development images are placed in the **Assets** folder. It is not a must-be location like Assets Catalogs in iOS or `drawable` on Android, but by convention most developer place image assets there. Therefore I suggest making this the default location for images instead of the project root. This could be done quite easily by modifying the `Xamarin.Forms.Platform.UWP.FileImageSourceHandler`. Although this is a breaking change, I think it is one worth making in the long run, as it makes the Xamarin.Forms development for UWP much more pleasant.

### (pseudo)Subfolders

Going further, it could be possible to add subfolder support for images as well. This would allow developers to organize their images into subfolders and refer to them as for example:

```
<Image Source="Icons/Help.png" />
```

This works in UWP, but the clear problem is that Android and iOS have the limitation of having images on one level (like all directly in the `drawable` folder) and subfolders are not allowed. For this I propose that the `/` characters in image paths would be replaced by `_` for these platforms - which means another simple modification in the platform-specific `FileImageSourceHandler` classes. Although this still means the images have to be in one single location, we can create a "subfolder-like" structure, which is easier to navigate. In addition - this is an optional feature - the developer who does not want to use this subfolder structure can still use normal filenames, even utilize underscores within them without a problem.

I have been using this approach in my Xamarin.Forms project quite successfully so far (utilizing a custom markup extension to perform the `/` to `_` replacement). Example result of this in the project folders:

**Android**
image

**UWP**
image

## Intended Use Case

The first part of this proposal - changing default image folder to `Assets` folder in UWP is a breaking change that will make the Xamarin.Forms UWP project root folders much less cluttered.

The pseudo-subfolder feature for images is an optional feature which allows developers to use `/` characters when referring to images in XAML, which is then evaluated to subfolders on UWP and underscore-based prefixes on Android and iOS.

### LINKED ISSUE #3228: Feature Request - Default Search Directory for UWP Icons (Platform Specific)
I would really like to put all of my icons/images inside of the `Assets` folder, however this creates some not so great looking code:

```
<Image>
                <Image.Source>
                    <OnPlatform x:TypeArguments="ImageSource">
                        <On Platform="iOS" Value="icon_twitter.png"/>
                        <On Platform="Android" Value="icon_twitter.png"/>
                        <On Platform="UWP" Value="Assets/icon_twitter.png"/>
                    </OnPlatform>
                </Image.Source>
            </Image>
```

Ideally I would like to do:

```xml
<Image Source="icon_twitter.png"/>
```

I could do this if I put it into the UWP root, but that clutters up the project if it is large.

I would like something like:

```
On<UWP>().SetImageSearchDirectory("Assets");
```

Then my nice pretty Image would work.

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

Added **Windows Platform Specific** to avoid set the default image directory.

Captura de pantalla 2019-10-10 a las 11 51 36

### Issues Resolved ###

- fixes #3228
- fixes #2235

### API Changes ###

Added:
 - Windows Application Platform Specific in Xamarin.Forms Core project.

Changed:
 - Changes in UWP ImageRenderer.

`Application.Current.On<Windows>().SetImageSearchDirectory("Assets");`

### Platforms Affected ###
- Core
- UWP

### PR Checklist ###

- [ ] Has automated tests 
- [ ] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.UAP/ImageRenderer.cs [hunk 1]
```diff
@@ -189,5 +195,39 @@ void UpdateAspect()
...
+				return;
+			}
+
+			var imageSearchDirectory = Application.Current.OnThisPlatform().GetImageSearchDirectory();
+
+			if(!string.IsNullOrEmpty(imageSearchDirectory))
+			{
+				var source = Element.Source;
+
+				if (source is FileImageSource)
```
Comment (jamesmontemagno) [inline Xamarin.Forms.Platform.UAP/ImageRenderer.cs]: Can we use some C# features here:  if (source is FileImageSource fileSource && fileSource != null)

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+﻿using Xamarin.Forms.CustomAttributes;
+using Xamarin.Forms.Internals;
+using Xamarin.Forms.PlatformConfiguration;
+using Xamarin.Forms.PlatformConfiguration.WindowsSpecific;
+
+namespace Xamarin.Forms.Controls.Issues
+{
+	[Preserve(AllMembers = true)]
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs]: insert `#if APP` here

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+			Application.Current.On<Windows>().SetImageSearchDirectory("Assets");
+		}
+
+		protected override void Init()
+		{
+
+		}
+	}
+}
```
Comment (samhouts) [inline Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs]: insert `#endif` here

File: Xamarin.Forms.Platform.UAP/ImageRenderer.cs [hunk 4]
```diff
@@ -189,5 +195,39 @@ void UpdateAspect()
...
+
+					if (fileSource != null)
+					{
+						var filePath = fileSource.File;
+
+						var directory = Path.GetDirectoryName(filePath);
+
+						if (string.IsNullOrEmpty(directory) || !Path.GetFullPath(directory).Equals(Path.GetFullPath(imageSearchDirectory)))
+						{
+							filePath = string.Format("{0}{1}{2}", imageSearchDirectory, Path.DirectorySeparatorChar, filePath);
```
Comment (samhouts) [inline Xamarin.Forms.Platform.UAP/ImageRenderer.cs]: Why not `Path.Combine`?

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs (hunk 2 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs]: Added.

File: Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs (hunk 3 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue3228.xaml.cs]: Added.

File: Xamarin.Forms.Platform.UAP/ImageRenderer.cs (hunk 4 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.UAP/ImageRenderer.cs]: There is no reason and in fact, it is better. Changed. Thanks!.

File: Xamarin.Forms.Platform.UAP/ImageRenderer.cs (hunk 1 above)
Comment (jsuarezruiz) [reply Xamarin.Forms.Platform.UAP/ImageRenderer.cs]: Sure,. changed.

**Comment by jsuarezruiz** [comment]
Ops, I did not see this [PR ](https://github.com/xamarin/Xamarin.Forms/pull/3263)and it's interesting. It's a good question. Maybe are compatible and both can be added. @davidortinau, what do you think?

**Comment by davidortinau** [comment]
Do we need to go with a platform specific? Could this be a config setting done at or near Xamarin.Forms.Init()?

@jassmith @samhouts

**Comment by jamesmontemagno** [comment]
That was my other idea too @davidortinau looking at it. I only ever need to set it once really IMHO.

**Comment by samhouts** [comment]
For @jamesmontemagno's request, a config setting is perfect. We can probably support @MartinZikmund's request for subfolders (#2335) in a different way altogether, closer to @paymicro's solution in #3263 but with a core property instead of a platform specific.

**Comment by VladislavAntonyuk** [comment]
Will it solve the same problem for wpf?

**Comment by dotMorten** [comment]
Why doesn't XF have something like WPF's pack URIs that works cross-plat and looks up the right way. Similarly we need a xplat build action that does the right thing per platform

**Comment by SamuelDebruyn** [comment]
This line in the ButtonRenderer also has to be modified: https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.UAP/ButtonRenderer.cs#L169

Otherwise the ButtonRenderer will still look in the root for images set as `Image` on a `Button`.

File: Xamarin.Forms.Platform.UAP/ImageRenderer.cs [hunk 5]
```diff
@@ -189,5 +195,34 @@ protected async Task UpdateSource()
...
+			{
+				var source = Element.Source;
+
+				if (source is FileImageSource fileSource && fileSource != null)
+				{
+					var filePath = fileSource.File;
+
+					var directory = Path.GetDirectoryName(filePath);
+
+					if (string.IsNullOrEmpty(directory) || !Path.GetFullPath(directory).Equals(Path.GetFullPath(imageSearchDirectory)))
```
Comment (kingces95) [inline Xamarin.Forms.Platform.UAP/ImageRenderer.cs]: So if the file appears to live at the search path then we don't combine the path with the search path. Means we can't do `assets\assets` right? Is there a reason to be backward compatible after someone has set the platform specific? Why not just drop this `if` and always combine? Certainly would be easier to document: 
> We prepend the search path to all asset paths.

**Comment by MartinZikmund** [comment]
Is this still considered? It would be great if it were part of Xamarin.Forms soon :-) .

**Comment by jsuarezruiz** [comment]
This PR has the same objective as: https://github.com/xamarin/Xamarin.Forms/pull/3263
Let's close one and focus on the other.

**Comment by jsuarezruiz** [comment]
Updated branch (resolved conflicts, rebase master, etc).

**Comment by samhouts** [comment]
@jsuarezruiz Please make sure this covered the same scenarios as #3263. Thanks!

**Comment by hartez** [comment]
I'm waaayyyy late to the party here, but the name of the property seems a little strange. It's not really a "search", it's just the location of the images. Something like `SetImageLocation` or `SetImageFolder` or `SetImageDirectory` seems more appropriate.

**Comment by jsuarezruiz** [comment]
@hartez Mnn, you're right that the name could perhaps confuse. I have renamed `ImageSearchDirectory` to `ImageDirectory`. Although, if there is another option that is more appropriate, we can still change it.

**Comment by jamesmontemagno** [comment]
WHOA!!!!!!! This is very exciting!!!!

**Pull request merged**